cypress/plugins/autorecord/index.ts

```diff
@@ -0,0 +1,201 @@
+import path from 'path'
+import { parse as parseUrl } from 'url'
+
+import { blobToPlain } from './util'
+
+const stringOrEmpty = (maybeString: any): string =>
+  typeof maybeString === 'string' ? maybeString : ''
+
+const stringArrayOrEmpty = (maybeArray: any): string[] =>
+  Array.isArray(maybeArray) && maybeArray.every(el => typeof el === 'string')
+    ? (maybeArray as string[])
+    : []
+
+const booleanOrFalse = (maybeBoolean: any): boolean =>
+  typeof maybeBoolean === 'boolean' ? maybeBoolean : false
+
+const cypressConfig = Cypress.env() || {}
+const { cleanMocks, forceRecord, recordTests, includeHosts, whitelistHeaders } = cypressConfig
+const shouldCleanMocks = booleanOrFalse(cleanMocks)
+const shouldForceRecord = booleanOrFalse(forceRecord)
+const testsToRecord = stringArrayOrEmpty(recordTests)
+const includedHosts = stringArrayOrEmpty(includeHosts)
+const whitelistedHeaders = stringArrayOrEmpty(whitelistHeaders)
+
+// Intercept all requests
+const interceptPattern = '*'
+
+const supportedMethods = ['get', 'GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'HEAD']
+
+const fileName =
+  stringOrEmpty(cypressConfig.mocksName) ||
+  path.basename(Cypress.spec.name, path.extname(Cypress.spec.name))
+const mocksFolder = stringOrEmpty(cypressConfig.mocksFolder) || 'cypress/mocks'
+
+before(function () {
+  if (shouldCleanMocks) {
+    cy.task('cleanMocks')
+  }
+
+  if (shouldForceRecord) {
+    // we don't save mocks to fixtures so we can skip this
+    // cy.task('removeAllMocks')
+  }
+})
+
+export function autoRecord() {
+  const whitelistHeaderRegexes = whitelistedHeaders.map((str: string) => RegExp(str))
+
+  // For cleaning, to store the test names that are active per file
+  const testNames: string[] = []
+  // For cleaning, to store the clean mocks per file
+  const cleanMockData: any = {}
+  // Locally stores all mock data for this spec file
+  let routesByTestId: any = {}
+  // For recording, stores data recorded from hitting the real endpoints
+  let routes: any[] = []
```

Comment:
```suggestion
  let routes: Record<string, unknown>[] = []
```

Reply:
Or even better, can we properly type this?

Even value has `{ url, method, status, data, body, headers }`, yeah?